File: swift/0535-encode-and-decode-tinyurl.swift
class Codec {
    var urlMap = [String: String]()

    // Encodes a URL to a shortened URL.
    func encode(_ longUrl: String) -> String {
        var id = UUID().uuidString

        if urlMap[id] != nil {
            id = UUID().uuidString
        }

        urlMap[id] = longUrl
        return "http://tinyurl.com/\(id)"
    }
    
    // Decodes a shortened URL to its original URL.
    func decode(_ shortUrl: String) -> String {
        let id = shortUrl.split(separator:"/").last!
        return urlMap[String(id)]!
    }
}

File: c++/0535-encode-and-decode-tinyurl.cpp
class Solution {
private:
    map<string, string> encodeMap;
    map<string, string> decodeMap;
    string base = "http://tinyurl.com/";
public:

    string encode(string longUrl) {
        if(!encodeMap.count(longUrl)) {
            string shortUrl = base + to_string(encodeMap.size() + 1);
            encodeMap[longUrl] = shortUrl;
            decodeMap[shortUrl] = longUrl;
        }
        return encodeMap[longUrl];
    }


    string decode(string shortUrl) {
        return decodeMap[shortUrl];
    }
};
